esx_service/utils/vmdk_utils.py

```diff
@@ -63,44 +63,55 @@
 VMDK_RETRY_COUNT = 5
 VMDK_RETRY_SLEEP = 1
 
-def init_datastoreCache():
+# For managing resource locks.
+lockManager = threadutils.LockManager()
+
+
+def init_datastoreCache(force=False):
     """
-    Initializes the datastore cache with the list of datastores accessible from local ESX host.
+    Initializes the datastore cache with the list of datastores accessible
+    from local ESX host. force=True will force it to ignore current cache
+    and force init
     """
     global datastores
-    logging.debug("init_datastoreCache: %s", datastores)
+    logging.debug("init_datastoreCache:  %s", datastores)
 
-    si = vmdk_ops.get_si()
+    with lockManager.get_lock("init_datastoreCache"):
+        if datastores and not force:
+            return
+
+        si = vmdk_ops.get_si()
 
-    #  We are connected to ESX so childEntity[0] is current DC/Host
-    ds_objects = \
-      si.content.rootFolder.childEntity[0].datastoreFolder.childEntity
-    datastores = []
+        #  We are connected to ESX so childEntity[0] is current DC/Host
+        ds_objects = si.content.rootFolder.childEntity[0].datastoreFolder.childEntity
+        tmp_ds = []
 
-    for datastore in ds_objects:
-        dockvols_path, err = vmdk_ops.get_vol_path(datastore.info.name)
-        if err:
-            continue
-        datastores.append((datastore.info.name,
+        for datastore in ds_objects:
+            dockvols_path, err = vmdk_ops.get_vol_path(datastore.info.name)
+            if err:
+                continue
```

Comment:
I'd suggest adding a logging.error here too, as it is not supposed to happen at all